### PR TITLE
feat(crds): Support scale selector. Fixes #5189

### LINKED ISSUE #5189: [Proposal] Support scale subresource in CRDs
# Summary

Support the `scale` subresource in the different Argo workflows CRDs

# Use Cases

Not having the `scale` subresource makes it difficult to use existing autoscalers like the vertical pod autoscaler. When running Argo in a multi-tenant environment, if a workflow definition does not have the appropriate resource requests and limits set, sudden spikes in CPU usage can cause pods to be evicted both for the workflow causing the spike and other tenants on the cluster. Having the `scale` subresource will allow autoscalers to spin up more nodes to handle any spikes.
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by JPZ13**
This PR:
- Addresses #5189 
- Adds the `scale` subresource that reports pods selected by a given workflow
- Adds an e2e test to confirm the selector output

The end result is that pods managed by Argo workflows can have their resources scaled up by the [vertical pod autoscaler](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler), helping to prevent pod evictions and failed workflow runs from resource usage spikes

Signed-off-by: J.P. Zivalich <j.p.zivalich@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

File: manifests/quick-start-postgres.yaml [hunk 1]
```diff
@@ -156,7 +156,12 @@ spec:
         type: object
     served: true
     storage: true
-    subresources: {}
+    subresources:
+      scale:
+        labelSelectorPath: .status.selector
+        specReplicasPath: .spec.parallelism
+        statusReplicasPath: .status.parallelism
+      status: {}
```
Comment (terrytangyuan) [inline manifests/quick-start-postgres.yaml:164]: Good to see this status subresource here. Would you like to update the PR title to reflect this change too?

Comment (JPZ13) [reply manifests/quick-start-postgres.yaml:164]: For sure! Thanks for the suggestion @terrytangyuan

Comment (JPZ13) [reply manifests/quick-start-postgres.yaml]: Including the status subresource seems to break everything. I'm going to remove it for now 😞

**Comment by alexec** [comment]
I think this is interesting, but I think that for more users, workflows are fire-and-forget. I'd like to hear from more users who want to scale a runtime.

**Comment by JPZ13** [comment]
> I think this is interesting, but I think that for more users, workflows are fire-and-forget. I'd like to hear from more users who want to scale a runtime.

Makes sense. I'm only really interested in the `selector` field so that our autoscaler can update resource requests and limits when they're not properly set without running into pod evictions. Would you like me to remove the `parallelism` field @alexec?

**Comment by alexec** [comment]
ok, so your use case is to scale up resource requests? I think it would be good to go narrow on that

**Comment by JPZ13** [comment]
> ok, so your use case is to scale up resource requests? I think it would be good to go narrow on that

Exactly. Adding the `selector` allows workflows to play nicely with the vertical pod autoscaler and prevents pod evictions. I'll remove the parallelism field and push another commit. Thanks @alexec

**Comment by JPZ13** [comment]
Unfortunately it seems that I can't remove the `replicas` field. It would still be nice to have access to the `selector` field though. Let me know how you would like me to proceed, and here's the relevant docs showing `replicas` as a required field: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource

**Comment by jessesuen** [comment]
I don't think using the scale subresource to adjust workflow parallelism is the right thing to be doing. Adjusting the parallelism of a workflow after it has started, is undefined behavior, and it's not something we want to support (which is what will happen with this feature). 

That said, the VPA use case is an interesting one. But as I understand, VPA deletes and recreates pods (as of its current implementation) when it scales. This will cause pods to become deleted from underneath the workflow controller, which will cause the workflow to fail. So this is not something a workflow can benefit from, at least until VPA is able to perform in-place updates of pods.

**Comment by JPZ13** [comment]
@jessesuen Thanks for pointing the VPA deletion/recreation out. I checked the [Known Limitations](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler#known-limitations) section and saw that updating running pods is only an experimental feature. 😞

Feel free to close this PR, and if you're ever doing some work on handling CPU spikes/pod evictions in a workflow, I'd be more than happy to contribute.

**Comment by jessesuen** [comment]
We can revisit this when VPA can do in-place updates.

**Pull request closed without merging**